runtime/src/utils/buffer/write.rs

```diff
@@ -53,64 +132,146 @@ impl<B: Blob> Write<B> {
     ///
     /// Panics if `capacity` is zero.
     pub fn new(blob: B, position: u64, capacity: usize) -> Self {
-        assert!(capacity > 0, "Buffer capacity must be greater than zero");
-        Self {
+        assert!(capacity > 0, "buffer capacity must be greater than zero");
+        let inner = Arc::new(RwLock::new(Inner {
             blob,
             buffer: Vec::with_capacity(capacity),
             position,
             capacity,
-        }
+        }));
+        Self { inner }
     }
+}
+
+impl<B: Blob> Blob for Write<B> {
```

Comment:
Unlike `buffer::Read`, the only interface we offer for `buffer::Write` now is a `Blob` (that makes all downstream code not have to worry about this at all).